### ISSUE TITLE
Agenda sometimes fails to initialize

### ISSUE THREAD

**Issue by kfrankovic** [agenda/agenda#510]
We have Node.js application hosted on Heroku which is using Agenda for the purpose of executing some scheduled tasks. But sometimes on dyno restart, Agenda simply fails to initialize (fire `ready `event, and consequentially it doesn't work). Roughly I would say it fails to initialize in around 50% of the restarts, and in that case we have to restart dyno several time to make it work. 

When it fails, there are no errors (we are subscribed to `error `event which we log, but there is nothing in log), and there are no other errors visible, so we are at dead end currently.

Scheduled tasks have their own separate dyno (to remove workload from main dynos), and dyno has it's own `procfile`. In that `procfile` we use `pm2` to start the process, as seen in next snippet (you'll see we have some extensive logging, just because we tried to catch the error): 

```
'use strict';

const pm2 = require('pm2');
const maxMemory = process.env.WEB_MEMORY || 512;

pm2.connect(() => {
	pm2.start({
		script: 'indexAgenda.js',
		name: 'WORKER',
		exec_mode: 'fork',
		instances: 1,
		max_memory_restart: maxMemory + 'M',
		env: {
			'NODE_ENV': 'production',
		},
	}, (err) => {
		if (err) {
			console.log(err);
			return console.error('[ WORKER ] Error while launching Job Worker!');
		}
		console.log('[ WORKER ] Job Worker started!');
		pm2.launchBus((err, bus) => {
			bus.on('log:out', (packet) => (console.log('[ WORKER ] ' + packet.data)));
			bus.on('log:err', (packet) => {
				console.error('[ WORKER ] Error: ' + packet.process.name);
				console.error('[ WORKER ] Error: ' + packet.data);
			});
		});
	});
});
```
PM2 will run our script. This script will run our transpiled and bundled javascript code:

```
if (process.env.NODE_ENV === 'production') {
  console.log('indexAgenda.js started');
  require('babel-polyfill');
  require('./distserver/agenda.bundle.js');
}
```
Here are most important parts of our agenda initialization (not bundled agenda.js file):

```
import agendaService from './services/AgendaService';
import logger from './util/logger';

if (process.env.NODE_ENV === 'production') {
  const mongoose = require('mongoose'); // eslint-disable-line global-require
  mongoose.Promise = require('bluebird'); // eslint-disable-line global-require
  require('./cloudwatchLogger')(); // eslint-disable-line global-require
  mongoose.connect(process.env.MONGOURL , (err) => {
    if (err) {
      console.error('Please make sure Mongodb is installed and running!'); // eslint-disable-line no-console
      logger.trackException(err, null, "Connecting agenda to the database failed.");
    }
    else {
      agendaService.initialize();

      // Log Agenda process exception
      process.on('uncaughtException', processErr => {
        try {
          logger.trackException(processErr, null, `AGENDA: Uncaught exception: ${processErr.message}`);
        }
        catch (e) {
          // eslint-disable-next-line no-console
          console.log("AGENDA: Tracking uncaught exception failed.");
          logger.trackException(e, null, `AGENDA: Tracking uncaught exception failed: ${e.message}`);
        }
        process.exit(1);
      });
    }
  });
}
else {
  agendaService.initialize();
}
```

AgendaService.js looks as follows (imports and big chunk of the code removed): 

```
const agenda = new Agenda({ db: { address: process.env.MONGOURL } }, (err) => {
  winstonLogger.logInfo("AGENDA: Agenda constructor called");
  if (err) {
    logger.trackException(err, null, "AGENDA: Agenda constructor error");
  }
});

function initialize() {
  console.log('Initializing AgendaService...'); // eslint-disable-line no-console
  winstonLogger.logInfo('Initializing AgendaService...');
  agenda.on('ready', () => {
    agenda.every(process.env.SCHEDULERFRENQUENCY, process.env.REMOVETOKENS);
    agenda.start();
    console.log('AgendaService initialized'); // eslint-disable-line no-console
    winstonLogger.logInfo('AgendaService initialized');
  });

  agenda.on('error', (err) => {
    logger.trackException(err, null, "AGENDA: Agenda error");
  });
}
```
No matter if it starts or fails, we always see `Initializing AgendaService...` message, and in case of successful start we see `AgendaService initialized`. But of course, if it doesn't start we don't see that message, there are no errors and it doesn't work. It looks like it is stuck. 

Did anyone experience any similar issues? Are we missing something?

**Comment by OmgImAlexis** [comment]
What version of Agenda are you on? Have you tried updating to the latest `1.0.0` version?

**Comment by kfrankovic** [comment]
We are using `0.9.1`. Are there any breaking changes in the `1.0.0`?

**Comment by simison** [comment]
@kfrankovic 1.0.0 has small breaking changes, see [History.md](https://github.com/agenda/agenda/blob/master/History.md).

You could update at least to 0.10.2 to to be able to enabled debugging, [see instructions](https://github.com/agenda/agenda#to-turn-on-logging-please-set-your-debug-env-variable-like-so). That might give us more insight into the problem.

**Comment by kfrankovic** [comment]
@simison Thanks for the suggestion. We updated Agenda to the 0.10.2, enabled logging and managed to provoke the issue with several restarts. There weren't any special errors visible in the log, it was simply - stuck.

The only different thing was that the order of execution was different. (visible from console logs and debug logs).

After some further investigation, we now suspect that in some cases in our code Agenda fires ready event **before** we had chance to attach ready handler. In ready handler, as can be visible in code I pasted at the top we have agenda.start(), and console.log for successful initialization. 

So in short, we currently believe that this looks like this reported issue: 

[https://github.com/agenda/agenda/issues/417](url)

... and that it wasn't stuck but fired ready event before we subscribed our listener.

**Comment by OmgImAlexis** [comment]
@kfrankovic we'd really appreciate if you could create an example script to reproduce this. I've been working on the tests for the last week and haven't noticed a single instance where ready fired before the handler was attached.

**Comment by kfrankovic** [comment]
@OmgImAlexis The code I pasted in my question is essentially what is sometimes producing the issue.

I can try to write something similar in following days that you will be able to run, though I'm pretty sure it won't reproduce error in consistent manner, since my assumption is that it depends on the time agenda package takes to fire ready event - and I think this is tightly coupled to the speed of database (I didn't had time to check the source code of agenda to get better insight).

**Comment by v1adko** [comment]
For those who'll be searching for similar issues, I had a problem where 'ready()' would not be called and seem to hang (no errors either), but in my case it was caused by the fact that I'm reusing mongoose connection (passing it as a mongo option) and the connection needs to be established (type - connected) before agenda init. Unfortunately agenda does not track that.

I've solved it by adding this before doing `new Agenda({ mongo })` just to avoid unexpected problems.
```javascript
const mongo = mongoose.connection;

if (mongo.states[mongo.readyState] === 'disconnected') {
    throw new Error('Something is wrong, DB connection not previously established');
}
// if (!mongo.readyState) is essentially the same thing, I just like the more explicit way
```

@OmgImAlexis Maybe agenda should do it internally?

**Comment by lazarz** [comment]
I have the same problem. "agenda@1.0.3",
Log in case of failed initialization
  agenda:database successful connection to MongoDB using collection: [agendaJobs] +0ms
  agenda:db_init init database collection using name [agendaJobs] +0ms
  agenda:db_init attempting index creation +1ms
  agenda:db_init index creation success +58ms

Log in case of successful initialization:
  agenda:database successful connection to MongoDB using collection: [agendaJobs] +0ms
  agenda:db_init init database collection using name [agendaJobs] +0ms
  agenda:db_init attempting index creation +1ms

  agenda:db_init index creation success +4ms
  agenda:define job [ping] defined with following options: 
  agenda:define { fn: [Function],
  agenda:define   concurrency: 5,
  agenda:define   lockLimit: 0,
  agenda:define   priority: 0,
  agenda:define   lockLifetime: 500,
  agenda:define   running: 0,
  agenda:define   locked: 0 } +0ms

  agenda:start Agenda.start called, creating interval to call processJobs every [5000ms] +0ms
  agenda:every Agenda.every(10 seconds, 'ping', [Object], undefined, cb) +0ms
  agenda:create Agenda.create(ping, [Object]) +0ms
  agenda:job [ping:undefined] computing next run via interval [10 seconds] +0ms
  agenda:saveJob attempting to save a job into Agenda instance +0ms
  agenda:saveJob set job props: 
  agenda:saveJob { name: 'ping',
  agenda:saveJob   data: undefined,
  agenda:saveJob   type: 'single',
  agenda:saveJob   priority: 0,
  agenda:saveJob   nextRunAt: 2018-06-27T19:57:06.425Z,
  agenda:saveJob   repeatInterval: '10 seconds',
  agenda:saveJob   repeatTimezone: null,
  agenda:saveJob   lastModifiedBy: undefined } +0ms
  agenda:saveJob current time stored as 2018-06-27T19:57:06.427Z +0ms
  agenda:saveJob job with type of "single" found +0ms
  agenda:saveJob job has a scheduled nextRunAt time, protecting that field from upsert +0ms
  agenda:saveJob calling findOneAndUpdate() with job name and type of "single" as query +0ms

  agenda:internal:processJobs starting to process jobs +0ms
  agenda:internal:processJobs queuing up job to process: [ping] +1ms
  agenda:internal:processJobs job [ping] lock status: shouldLock = true +0ms
  agenda:_findAndLockNextJob _findAndLockNextJob(ping, [Function], cb) +0ms

**Comment by johnboscopyra** [comment]
https://stackoverflow.com/questions/53460679/unhandledpromiserejectionwarning-when-using-agenda/63007932#63007932

**Comment by nareshsr** [comment]
https://stackoverflow.com/questions/53460679/unhandledpromiserejectionwarning-when-using-agenda/63007932#63007932
